### DISCUSSION TITLE
Why are `Sink`s responsible for maintaining statistics, when isn't part of the `Sink` trait definition?

### DISCUSSION THREAD

**Discussion by BatmanAoD** [BurntSushi/ripgrep#2271] [Q&A] [answered]
I'm trying to understand how the printers are used, and I'm looking at `search_path` and `search_reader` in `search.rs`. For both search functions and for all three types of printers, `sink.stats()` and `sink.has_match()` are called to populate the fields of the `SearchResult` object,

But `stats()` and `has_match()` are not required by the `Sink` trait. If sinks are intended to be uniformly responsible for maintaining the search statistics, why not add those functions to the definition of `Sink`? (Of course, `JSONSink.stats()` would need to be changed to return an option like the other `stats` implementations do.) Conversely, if there's intended to be some flexibility around whether or not the `Sink` trait implements these functions, why not add a separate trait (something like `StatsSink` or `StatsRecorder`) to codify this?

**Comment by BurntSushi** [comment, marked as answer]
> why not add a separate trait

When it comes to generics, I _never_ ask "why not." I always ask "why do it in the first place."

Stats don't seem like a fundamental part of "sink" to me. It's true that all of the printers provide stats, but note that not all `Sink` impls provide stats.

Sorry if this isn't a fulfilling answer, but it really is that simple: IMO you have the burden of proof flipped. :-)

**Reply by BatmanAoD** [reply]
No, that's fair; I'm approaching this in complete ignorance of the design, and not a lot of concrete Rust experience!

The thing in the code that made me think `has_match` and `stats` should be codified as part of a trait is the code duplication on the match-arms in `search_path` and `search_reader`. I realize that there are only three arms in each one, and only a couple of lines in each arm, but it does seem a little odd to need to continue that duplication if there are ever more printers added.

**Reply by BurntSushi** [reply]
It would be very helpful if you used github permalinks to the code you're talking about. It makes your questions a lot more accessible. It would potentially let me answer it in mobile for example.

**Reply by BatmanAoD** [reply]
@BurntSushi Oh, sure, sorry.

Here are `search_path` and `search_reader`: https://github.com/BurntSushi/ripgrep/blob/master/crates/core/search.rs#L474

**Reply by BurntSushi** [reply]
Ah I see. That level of code duplication doesn't even register for me. It's trivial and localized.

Remember, DRY is a _means_, not an _end_. :-)
